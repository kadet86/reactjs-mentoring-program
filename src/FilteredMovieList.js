import React from 'react';
import MovieList from './MovieList';
import SearchForm from './SearchForm';
import SearchResultsInfo from './SearchResultsInfo';

const URL = 'https://reactjs-cdp.herokuapp.com/movies';

export default class FilteredMovieList extends React.PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            movies: [],
            query: '',
            searchBy: 'title',
            sortBy: 'release_date'
        };
    }

    fetchMovies() {
        const {query, searchBy, sortBy} = this.state;

        fetch(`${URL}?sortOrder=desc&limit=50&search=${query}&searchBy=${searchBy}&sortBy=${sortBy}`)
            .then(res => res.json())
            .then(json => {
                this.setState({movies: json.data});
            });
    }

    componentDidMount() {
        this.fetchMovies();
    }

    onQueryChange = (event) => {
        this.setState({query: event.target.value});
    }

    onSortByChange = (event) => {
        this.setState({sortBy: event.target.getAttribute('data-value')}, () => {
            this.fetchMovies();
        });
    }

    onSearchByChange = (event) => {
        event.preventDefault();
        let target = event.target;
        if (target.tagName !== 'BUTTON') {
            target = target.parentNode;
        }
        this.setState({searchBy: target.getAttribute('data-value')});
    }

    onSubmit = (event) => {
        event.preventDefault();
        this.fetchMovies();
    }

    render() {
        return (
            <section className="filtered-movie-list">
                <div className="top-section">
                    <div>
                    <h1>FIND YOUR MOVIE</h1>
                    <SearchForm 
                        query={this.state.query} 
                        searchBy={this.state.searchBy} 
                        onSearchByChange={this.onSearchByChange}
                        onQueryChange={this.onQueryChange}
                        onSubmit={this.onSubmit} />
                    </div>
                </div>
                <SearchResultsInfo 
                    resultsCount={this.state.movies.length}
                    sortBy={this.state.sortBy} 
                    onSortByChange={this.onSortByChange} />
                <MovieList 
                    movies={this.state.movies} 
                    navigateToMovie={this.props.navigateToMovie} />
            </section>
        );
    }
}