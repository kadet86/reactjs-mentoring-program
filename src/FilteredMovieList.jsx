// @flow
import * as React from 'react';
import Router from 'next/router';
import { Toolbar } from 'primereact/toolbar';
import { connect } from 'react-redux';
import {
    changeQuery,
    changeSearchBy,
    changeSortBy,
    getMovies,
    showMovies,
} from './actions';
import MovieList from './MovieList';
import SearchForm from './SearchForm';
import SearchResultsInfo from './SearchResultsInfo';
import TopSection from './TopSection';
import type { MovieInfo } from './MovieListItem';
import { createSelector } from 'reselect';
import styled from 'styled-components';
export const MOVIES_LIMIT = 50;

export function buildSearchRoute({
    query,
    searchBy,
    sortBy,
}: {
    query: string,
    searchBy: string,
    sortBy: string,
}) {
    if (!query) {
        return {
            url: '/search',
            as: '/search',
        };
    }

    query = encodeURIComponent(query);
    return {
        url: `/search?query=${query}&searchBy=${searchBy}&sortBy=${sortBy}`,
        as: `/search/${query}/${searchBy}/${sortBy}`,
    };
}

type RouterInfo = {
    asPath: string,
    query: {
        query?: string,
        searchBy?: string,
        sortBy?: string,
    },
};

type Props = {
    query: string,
    searchBy: string,
    sortBy: string,
    movies: any[],
    router: RouterInfo,
    showMovies: (movies: MovieInfo[]) => void,
    getMovies: ({
        limit: number,
        query?: string,
        searchBy?: string,
        sortBy?: string,
    }) => void,
    changeQuery: (query?: string) => void,
    changeSearchBy: (searchBy?: string | null) => void,
    changeSortBy: (sortBy?: string | null) => void,
};

const Section = styled.section`
    width: 100%;
`;

const CustomToolbar = styled(Toolbar)`
    height: 40px;
`;

const Header = styled.h1`
    margin: 0;
    margin-bottom: 10px;
    font-size: 14px;
    color: white;
`;

const NoResultsSection = styled.section`
    height: 300px;
    display: flex;
    align-items: center;
    justify-content: center;
`;

const NoResultsSectionHeader = styled.h1`
    color: black;
    font-size: 36px;
`;

export class FilteredMovieList extends React.PureComponent<Props> {
    constructor(props: Props) {
        super(props);

        this.onQueryChange = this.onQueryChange.bind(this);
        this.onSearchByChange = this.onSearchByChange.bind(this);
        this.onSortByChange = this.onSortByChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.syncParams();
    }

    syncParams() {
        const { query, searchBy, sortBy } = this.props.router.query;

        if (query !== this.props.query) {
            this.props.changeQuery(query);
        }

        if (searchBy && searchBy !== this.props.searchBy) {
            this.props.changeSearchBy(searchBy);
        }

        if (sortBy && sortBy !== this.props.sortBy) {
            this.props.changeSortBy(sortBy);
        }
    }

    componentDidMount() {
        this.fetchMovies();
    }

    componentDidUpdate(prevProps: Props) {
        if (prevProps.router.asPath !== this.props.router.asPath) {
            this.syncParams();
            this.fetchMovies();
        }
    }

    navigate(params: any = {}) {
        const { url, as } = buildSearchRoute({
            ...this.props,
            ...params,
        });
        Router.push(url, as);
    }

    fetchMovies() {
        const { query, searchBy, sortBy } = this.props.router.query;
        if (!query) {
            this.props.showMovies([]);
            return;
        }

        this.props.getMovies({
            limit: MOVIES_LIMIT,
            query,
            searchBy,
            sortBy,
        });
    }

    /*:: onQueryChange: () => void */
    onQueryChange(event: SyntheticEvent<HTMLInputElement>): void {
        this.props.changeQuery(event.currentTarget.value);
    }

    /*:: onSortByChange: () => void */
    onSortByChange(event: SyntheticEvent<HTMLElement>): void {
        const sortBy = event.currentTarget.getAttribute('data-value');
        this.props.changeSortBy(sortBy);
        this.navigate({ sortBy });
    }

    /*:: onSearchByChange: () => void */
    onSearchByChange(event: SyntheticEvent<HTMLElement>): void {
        event.preventDefault();
        this.props.changeSearchBy(
            event.currentTarget.getAttribute('data-value')
        );
    }

    /*:: onSubmit: () => void */
    onSubmit(event: SyntheticEvent<HTMLElement>): void {
        event.preventDefault();
        this.navigate();
    }

    render() {
        const noResults = !this.props.movies || this.props.movies.length === 0;
        return (
            <Section>
                <TopSection>
                    <Header>FIND YOUR MOVIE</Header>
                    <SearchForm
                        query={this.props.query}
                        searchBy={this.props.searchBy}
                        onSearchByChange={this.onSearchByChange}
                        onQueryChange={this.onQueryChange}
                        onSubmit={this.onSubmit}
                    />
                </TopSection>
                {!noResults && (
                    <>
                        <SearchResultsInfo
                            resultsCount={this.props.movies.length}
                            sortBy={this.props.sortBy}
                            onSortByChange={this.onSortByChange}
                        />
                        <MovieList movies={this.props.movies} />
                    </>
                )}
                {noResults && (
                    <>
                        <Toolbar />
                        <NoResultsSection>
                            <NoResultsSectionHeader>
                                No films found
                            </NoResultsSectionHeader>
                        </NoResultsSection>
                    </>
                )}
            </Section>
        );
    }
}

const getStateMovies = state => state.movies;
const moviesSelector = createSelector(
    getStateMovies,
    movies => movies
);
const makeMapStateToProps = () => {
    const mapStateToProps = state => {
        return {
            query: state.query,
            sortBy: state.sortBy,
            searchBy: state.searchBy,
            movies: moviesSelector(state),
        };
    };
    return mapStateToProps;
};

const mapDispatchToProps = {
    getMovies,
    changeSortBy,
    changeQuery,
    changeSearchBy,
    showMovies,
};

export default connect(
    makeMapStateToProps,
    mapDispatchToProps
)(FilteredMovieList);
