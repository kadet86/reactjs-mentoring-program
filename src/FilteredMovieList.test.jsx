import React from 'react';
import {shallow} from 'enzyme';
import FilteredMovieList from './FilteredMovieList';

describe('FilteredMovieList component', () => {
    beforeEach(() => {
        fetch.resetMocks();
    });

    it('has persistent default snapshot', () => {
        fetch.mockResponseOnce(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);
        expect(component).toMatchSnapshot();
    });

    it('updates state.query when onQueryChange() is called', () => {
        fetch.mockResponseOnce(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);

        expect(component.state('query')).toBe('');
        
        component.instance().onQueryChange({target: {value:'new query'}});

        expect(component.state('query')).toBe('new query');
    });

    it('updates state.searchBy when onSearchByChange() is called', () => {
        fetch.mockResponseOnce(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);

        expect(component.state('searchBy')).toBe('title');
        
        component.instance().onSearchByChange({
            preventDefault: () => {}, 
            target: {tagName: 'BUTTON', getAttribute: () => 'genres'},
        });

        expect(component.state('searchBy')).toBe('genres');
    });

    it('updates state.searchBy when onSearchByChange() is called (span click)', () => {
        fetch.mockResponseOnce(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);

        expect(component.state('searchBy')).toBe('title');
        
        component.instance().onSearchByChange({
            preventDefault: () => {}, 
            target: {
                tagName: 'SPAN', 
                parentNode: {getAttribute: () => 'genres'},
            },
        });

        expect(component.state('searchBy')).toBe('genres');
    });

    it('updates state.sortBy and fetches moview when onSortByChange() is called', () => {
        fetch.mockResponse(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);

        expect(component.state('sortBy')).toBe('release_date');
        
        component.instance().onSortByChange({
            preventDefault: () => {}, 
            target: {tagName: 'BUTTON', getAttribute: () => 'rating'},
        });

        expect(component.state('sortBy')).toBe('rating');
        expect(fetch.mock.calls[1][0]).toEqual('https://reactjs-cdp.herokuapp.com/movies?sortOrder=desc&limit=50&search=&searchBy=title&sortBy=rating');
    });

    it('fetches movies when onSubmit() is called', () => {
        fetch.mockResponse(JSON.stringify({ data: [] }));
        const component = shallow(<FilteredMovieList />);
         
        component.instance().onSubmit({
            preventDefault: () => {},
        });

        expect(fetch.mock.calls[1][0]).toEqual('https://reactjs-cdp.herokuapp.com/movies?sortOrder=desc&limit=50&search=&searchBy=title&sortBy=release_date');
    });
});
