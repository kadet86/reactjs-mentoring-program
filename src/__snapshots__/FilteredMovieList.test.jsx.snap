// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredMovieList component has persistent default snapshot 1`] = `ShallowWrapper {}`;
